engine/verification/finder/engine.go

```diff
@@ -61,12 +66,14 @@ func New(
 		log:                      log.With().Str("engine", "finder").Logger(),
 		metrics:                  metrics,
 		me:                       me,
+		state:                    state,
 		match:                    match,
 		headerStorage:            headerStorage,
 		cachedReceipts:           cachedReceipts,
 		pendingReceipts:          pendingReceipts,
 		readyReceipts:            readyReceipts,
 		processedResultIDs:       processedResultIDs,
+		discardedResultIDs:       discardedResultIDs,
```

Comment:
I suggest we remove this mempool and check the state directly as needed with `stakedAtBlockID`. I believe this mempool practically acts as a cache for the result of `stakedAtBlockID`, which is essentially a `snapshot.Identity` call. In general I think we should avoid adding extra caches and optimizations like this unless there is a clear need for it. In particular here I feel that our energy would be better spent optimizing the underlying `Identity` call when that becomes necessary rather than optimizing this layer.

Reply:
Per our conversation, we use this mempool to avoid passing an already discarded receipt down the pipeline, as it saves at least one mempool read and write per duplicate discarded receipt. As reading and writing with mempools require taking the identity of the entity, this extra read and write would take unnecessary cpu utilization per earlier profiling. 

Also, after carefully reading the code and considering concurrent execution of pipeline components (i.e., cached, pending, and ready), it seems that we need to check for discarded/processed receipts at 2 places:
1- When adding receipts to cached mempool.
2- When adding receipts to ready mempool. 

I think the first case is clear (per our conversation), however, regarding the second case, let's assume the following example:
- Receipts A and B both have the same result.
- Their result has been neither processed nor discarded. 
- Receipt A comes first and gets to ready mempool (i.e., cached --> ready) 
- Receipt A is _not processed yet_. 
- Receipt B comes next and gets to cached mempool. 
- Receipt A is processed. 
- Receipt B is moved from cached to ready, which should not because its result processed.

A similar case stands for discarding results. In general, since our ready mempool is of limited size, adding already processed or discarded results to it may result in ejecting valid results (i.e., those that are not processed yet). 

Accordingly, I have added the checks at both places.

Reply:
I'm fine to keep discarded ids separately. 

Now the receipt has a few states, I would suggest adding comments somewhere to explain them:

- cached: the receipt is received but not handled yet
- pending: the receipt is handled, but the block is not received yet
- discarded: the receipt's block is received, but we learned we are not staked at this block
- ready: the receipt's block is received, and we are staked at this block, it's ready to be forwarded to match engine
- processed: the receipt has been forwarded to matching engine.

Reply:
addressed in https://github.com/onflow/flow-go/pull/245/commits/009b7114a0acc32db46c1999a9626925010cbb8a